tests/test_non_linearity_correction.py

```diff
@@ -0,0 +1,154 @@
+
+        Both of these interpolations are linear, and both use their edge values as
+        constant extrapolations for out of bounds values.
+    """
+
+    # Get file data
+    gain_ax, count_ax, relgains = _parse_file(nonlin_path)
+
+    # Create interpolation for em gain (x), counts (y), and relative gain (z).
+    # Note that this defaults to using the edge values as fill_value for
+    # out of bounds values (same as specified below in interp1d)
+    f = interpolate.RectBivariateSpline(gain_ax,
+                                    count_ax,
+                                    relgains.T,
+                                    kx=1,
+                                    ky=1,
+    )
+    # Get the relative gain curve for the given gain value
+    relgain_curve = f(em_gain, count_ax)[0]
+
+    # Create interpolation for dn counts (x) and relative gains (y). For
+    # out of bounds values use edge values
+    ff = interpolate.interp1d(count_ax, relgain_curve, kind='linear',
+                              bounds_error=False,
+                              fill_value=(relgain_curve[0], relgain_curve[-1]))
+    # For each dn count, find the relative gain
+    counts_flat = ff(frame.ravel())
+
+    return counts_flat.reshape(frame.shape)
+
+def test_non_linearity_correction():
+    """
+    Generate a non-linearity correction calibration and test the correction 
+    
+    Ported from II&T Pipeline
+    """
+    ###### create a simulated dataset that is non-linear
+    # check that simulated data folder exists, and create if not
+    datadir = os.path.join(os.path.dirname(__file__), "simdata")
+    if not os.path.exists(datadir):
+        os.mkdir(datadir)
+    
+    emgain = 2000
+    mocks.create_nonlinear_dataset(filedir=datadir,em_gain=emgain)
+
+    ####### open up the files
+    sim_data_filenames = glob.glob(os.path.join(datadir, "simcal_nonlin*.fits"))
+    nonlinear_dataset = data.Dataset(sim_data_filenames)
+    assert len(nonlinear_dataset) == 2
+
+    ######## perform non-linearity correction
+    non_linearity_correction = data.NonLinearityCalibration(os.path.join(os.path.dirname(__file__),'nonlin_sample.fits'))
+    linear_dataset = l1_to_l2a.correct_nonlinearity(nonlinear_dataset, non_linearity_correction)
+
+    #The data was generated with a ramp in the x-direction going from 10 to 65536
+    expected_ramp = np.linspace(10,65536,1024)
+    #Let's collapse the data and see if there's a ramp. 
+    collapsed_data = np.mean(linear_dataset.all_data, axis=(0,1))
+
+    #Relative correction
+    relative_correction = (collapsed_data-expected_ramp)/collapsed_data
+
+    #We are happy if the relative correction is less than 1% [TBC]
+    assert np.all(relative_correction < 1e-2)
+
+    
+    #Let's test that this returns the same thing as the II&T pipeline
+    linear_data_iit = nonlinear_dataset.all_data*get_relgains(nonlinear_dataset.all_data,emgain,os.path.dirname(os.path.abspath(__file__))+"/test_data/nonlin_sample.csv")
```

Comment:
best practice to use os.path.join() rather than adding strings for filenames: e.g., `os.path.join(os.path.dirname(os.path.abspath(__file__)), "test_data", "nonlin_sample.csv")`

Reply:
Done